index.ts

```diff
@@ -87,53 +84,117 @@ function getPackageVersion(pkg: PackageJson): string {
   return getString(pkg, 'version', `${getPackageName(pkg)}'s package.json is missing a version.`);
 }
 
-interface ProjectConstructor {
-  new (name: string, version?: string, cb?: (project: Project) => void, root?: string): Project;
-  fromJSON(json: fixturify.DirJSON, name: string): Project;
-  fromDir(root: string, name: string): Project;
+interface ReadDirOpts {
+  linkDeps?: boolean;
+}
+
+// This only shallow-merges with any user-provided files, which is OK right now
+// because this is only one level deep. If we ever make it deeper, we'll need to
+// switch to a proper deep merge.
+const defaultFiles = {
+  'index.js': `
+    'use strict';
+     module.exports = {};`,
+};
+
+export interface ProjectArgs {
+  name?: string;
+  version?: string;
+  files?: fixturify.DirJSON;
+  requestedRange?: string;
 }
 
-class Project {
+export class Project {
   pkg: PackageJson;
-  files: fixturify.DirJSON = {
-    'index.js': `
-'use strict';
-module.exports = {};`,
-  };
+  files: fixturify.DirJSON;
   readonly isDependency = true;
 
   private _dependencies: { [name: string]: Project } = {};
   private _devDependencies: { [name: string]: Project } = {};
-  private _root: string;
+  private _baseDir: string | undefined;
   private _tmp: tmp.SynchrounousResult | undefined;
 
-  constructor(name: string, version = '0.0.0', cb?: (project: Project) => void, root?: string) {
-    this.pkg = {
-      name,
-      version,
-      keywords: [],
-    };
+  // when used as a dependency in another Project, this is the semver range it
+  // will appear as within the parent's package.json
+  private requestedRange: string;
+
+  private dependencyLinks: Map<string, { dir: string; requestedRange: string }> = new Map();
+  private linkIsDevDependency: Set<string> = new Set();
+
+  constructor(name?: string, version?: string, args?: Omit<ProjectArgs, 'name' | 'version'>);
+  constructor(name?: string, args?: Omit<ProjectArgs, 'name'>);
+  constructor(args?: ProjectArgs);
+  constructor(
```

Comment:
So this constructor changed a-lot, specifically the trailing `cb`. Whats the rationale behind that?  At first glance this seems like unnecessary breaking change.

```
cb?: (project: Project) => void, root?: string
```

Reply:
(Note that everything I'm saying about this constructor signature applies as well to `addDependency` and `addDevDependency`, which should now be completely consistent with the constructor.)

The main problem I was fixing here was that the use of positional arguments was getting very awkward, especially since I was adding two new ones (`files` and `requestedRange`). 

After reviewing a lot of existing tests, it was clear that `version` for example was almost always being set not because the tests cared about the version, but only because it was positionally in the way of other arguments.

So my goal here is to support entirely named arguments instead. But to ease the upgrade, we also optionally support `name` and `version` as leading positional arguments.

Another thing I encountered while pairing and reviewing existing tests is that the use of the `cb` style was unnecessarily confusing. It's entirely redundant because it gets access to the same value returned by the constructor (or by `addDependency`, etc). Reducing the deep nesting made the tests much better. And keep `cb` around while also supporting the optional positional args and all the named args was making the implementation and type signature more complex than I liked, so I jettisoned it.

The most common thing I saw tests doing inside `cb` was setting `files`, which is now a thing you can pass directly as an argument instead.

Reply:
FWIW I actually agree with both of @ef4's points. The use of positional args definitely becomes extremely unruly when attempting to widen the signature. Additionally, I've always found the cb API confusing to use, as it behaves in a fairly unpredictable way in terms of scoping.